pkg/logqlmodel/metadata/context.go

```diff
@@ -0,0 +1,82 @@
+/*
+Package metadata provides primitives for recording metadata across the query path.
+Metadata is passed through the query context.
+*/
+package metadata
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase/definitions"
+)
+
+type (
+	ctxKeyType string
+)
+
+const (
+	metadataKey ctxKeyType = "metadata"
+)
+
+// Context is the metadata context. It is passed through the query path and accumulates metadata.
+type Context struct {
+	mtx     sync.RWMutex
+	headers map[string][]string
+}
+
+// NewContext creates a new metadata context
+func NewContext(ctx context.Context) (*Context, context.Context) {
+	contextData := &Context{
+		headers: map[string][]string{},
+	}
+	ctx = context.WithValue(ctx, metadataKey, contextData)
+	return contextData, ctx
+}
+
+// FromContext returns the metadata context.
+func FromContext(ctx context.Context) *Context {
+	v, ok := ctx.Value(metadataKey).(*Context)
+	if !ok {
+		return &Context{
+			headers: map[string][]string{},
+		}
+	}
+	return v
+}
+
+// Headers returns the cache headers accumulated in the context so far.
+func (c *Context) Headers() []*definitions.PrometheusResponseHeader {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	headers := make([]*definitions.PrometheusResponseHeader, 0, len(c.headers))
+	for k, vs := range c.headers {
+		header := definitions.PrometheusResponseHeader{
+			Name:   k,
+			Values: vs,
+		}
+		headers = append(headers, &header)
+	}
+
+	sort.Slice(headers, func(i, j int) bool {
```

Comment:
what's the reason for sorting the headers?

Reply:
It keeps the test the test deterministic 🙂